File: packages/tf-lang-l0-ts/src/model/bytecode.ts

export interface Program {
  version: string;
  regs: number;
  instrs: Instr[];
}

export type Instr =
  | { op: 'HALT' }
  | { op: 'CONST', dst: number, value: unknown }
  | { op: 'PACK', dst: number, tag: string, regs: number[] }
  | { op: 'UNPACK', dsts: number[], tag: string, src: number }
  | { op: 'ID_HASH', dst: number, src: number }
  | { op: 'SNAP_MAKE', dst: number, state: number }
  | { op: 'SNAP_ID', dst: number, snapshot: number }
  | { op: 'LENS_PROJ', dst: number, state: number, region: string }
  | { op: 'LENS_MERGE', dst: number, state: number, region: string, sub: number }
  | { op: 'PLAN_SIM', dst_delta: number, dst_world: number, world: number, plan: number }
  | { op: 'DIFF_APPLY', dst: number, state: number, delta: number }
  | { op: 'DIFF_INVERT', dst: number, delta: number }
  | { op: 'JOURNAL_REC', dst: number, plan: number, delta: number, snap0: number, snap1: number, meta: number }
  | { op: 'JOURNAL_REW', dst: number, world: number, entry: number }
  | { op: 'CALL', dst: number, tf_id: string, args: number[] }
  | { op: 'ASSERT', pred: number, msg: string };

File: packages/tf-lang-l0-rs/src/model/bytecode.rs

use serde::{Deserialize, Serialize};

/// SSA-style program; regs are numbered 0..regs-1
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Program {
    pub version: String,
    pub regs: u16,
    pub instrs: Vec<Instr>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Instr {
    Assert { pred: u16, msg: String },
    Halt,
    Const { dst: u16, value: serde_json::Value },
    Pack  { dst: u16, tag: String, regs: Vec<u16> },
    Unpack{ dsts: Vec<u16>, tag: String, src: u16 },

    IdHash { dst: u16, src: u16 },
    SnapMake { dst: u16, state: u16 },
    SnapId { dst: u16, snapshot: u16 },

    LensProj { dst: u16, state: u16, region: String },
    LensMerge{ dst: u16, state: u16, region: String, sub: u16 },

    PlanSim { dst_delta: u16, dst_world: u16, world: u16, plan: u16 },
    DiffApply { dst: u16, state: u16, delta: u16 },
    DiffInvert{ dst: u16, delta: u16 },

    JournalRec { dst: u16, plan: u16, delta: u16, snap0: u16, snap1: u16, meta: u16 },
    JournalRew { dst: u16, world: u16, entry: u16 },

    Call { dst: u16, tf_id: String, args: Vec<u16> },
}
